File: mobile/lib/ui/core/lifecycle_observer.dart
import 'package:flutter/material.dart';

import 'connection_coordinator.dart';

/// Watches OS-level lifecycle events and pauses/resumes the session through
/// the [ConnectionCoordinator].
///
/// Register this as a [WidgetsBindingObserver] during the driving view's
/// [State.initState] and remove it in [State.dispose].
class LifecycleObserver extends WidgetsBindingObserver {
  LifecycleObserver({required this.coordinator});

  final ConnectionCoordinator coordinator;

  @override
  void didChangeAppLifecycleState(AppLifecycleState state) {
    switch (state) {
      case AppLifecycleState.inactive:
      case AppLifecycleState.paused:
      case AppLifecycleState.detached:
        coordinator.pause();
      case AppLifecycleState.resumed:
        coordinator.resume();
      case AppLifecycleState.hidden:
        break;
    }
  }
}

File: android/app/src/main/kotlin/dev/fazrigading/wheeldeck/ui/core/LifecycleObserver.kt
package dev.fazrigading.wheeldeck.ui.core

import androidx.lifecycle.DefaultLifecycleObserver
import androidx.lifecycle.LifecycleOwner
import dev.fazrigading.wheeldeck.ui.features.connection.view_models.ConnectionViewModel

/// Watches OS-level lifecycle events and pauses/resumes the session through
/// the [ConnectionViewModel].
class LifecycleObserver(private val viewModel: ConnectionViewModel) : DefaultLifecycleObserver {
    override fun onPause(owner: LifecycleOwner) {
        viewModel.pause()
    }

    override fun onResume(owner: LifecycleOwner) {
        viewModel.resume()
    }
}
